ruby/ql/test/query-tests/security/cwe-312/logging.rb

```diff
@@ -0,0 +1,84 @@
+stdout_logger = Logger.new STDOUT
+
+password = "043697b96909e03ca907599d6420555f"
+
+# BAD: password logged as plaintext
+stdout_logger.info password
+# BAD: password logged as plaintext
+stdout_logger.debug password
+# BAD: password logged as plaintext
+stdout_logger.error password
+# BAD: password logged as plaintext
+stdout_logger.fatal password
+# BAD: password logged as plaintext
+stdout_logger.unknown password
+# BAD: password logged as plaintext
+stdout_logger.warn password
+
+# BAD: password logged as plaintext
+stdout_logger.add Logger::WARN, password
+# BAD: password logged as plaintext
+stdout_logger.add Logger::WARN, "message", password
+# BAD: password logged as plaintext
+stdout_logger.log Logger::WARN, password
+
+# BAD: password logged as plaintext
+stdout_logger << "pw: #{password}"
+# BAD: sensitive data in the progname will taint subsequent logging calls
+stdout_logger.progname = password
+
+hsh1 = { password: "aec5058e61f7f122998b1a30ee2c66b6" }
+hsh2 = {}
+# GOOD: no backwards flow
+stdout_logger.info hsh2[:password]
+hsh2[:password] = "beeda625d7306b45784d91ea0336e201"
+
+# BAD: password logged as plaintext
+stdout_logger.info hsh1[:password]
```

Comment:
Unless I've missed something, I don't think there's a result in the `.expected` file for this case.

Reply:
I must have missed this - Tom's suggested CFG node matching appears to fix this.